src/objects/utils/serializers.py

```diff
@@ -1,69 +1,125 @@
-from django.utils.translation import ugettext_lazy as _
+import logging
+from collections import defaultdict
 
+from glom import GlomError, glom
 from rest_framework import serializers
 
+from objects.token.constants import PermissionModes
+
+logger = logging.getLogger(__name__)
+
+
+def build_spec(fields) -> dict:
+    spec = {}
+    for spec_field in fields:
+        build_spec_field(spec, name=spec_field, value=spec_field)
+    return spec
+
+
+def build_spec_field(spec, name, value):
+    if "__" in name:
+        parent, field_name = name.split("__", 1)
+        spec[parent] = spec.get(parent, {})
+        build_spec_field(spec[parent], field_name, value)
+    else:
+        spec[name] = value.replace("__", ".")
+
+
+def get_field_names(data: dict) -> list:
+    field_names = []
+    for key, value in data.items():
+        if not isinstance(value, dict):
+            field_names.append(key)
+        else:
+            field_names += [f"{key}__{val}" for val in get_field_names(value)]
+
+    return field_names
+
+
+class NotAllowedDict(defaultdict):
+    def pretty(self):
+        if len(self.keys()) == 0:
+            return ""
+        if len(self.keys()) == 1:
+            value = list(self.values())[0]
+            return ",".join(value)
+        return "; ".join([f"{key}={','.join(value)}" for (key, value) in self.items()])
+
 
 class DynamicFieldsMixin:
     """
     this mixin allows selecting fields for serializer in the query param
-    It also supports nested fields which are serializers themselves.
+    It also supports nested fields.
     """
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
+        self.not_allowed = NotAllowedDict(set)
+
+    def to_representation(self, instance):
+        data = super().to_representation(instance)
+
+        allowed_fields = self.get_allowed_fields(instance)
         query_fields = self.get_query_fields()
+
+        if not query_fields and allowed_fields == ["*"]:
```

Comment:
Make "*" a constant somewhere to prevent magic strings.

Reply:
done